src/serialization/serialization.rs

```diff
@@ -0,0 +1,412 @@
+extern crate byteorder;
+
+use super::{Counter, Histogram};
+use std::io::{self, Cursor, Read, Write};
+use std;
+use super::num::ToPrimitive;
+use self::byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+
+#[path = "tests.rs"]
+#[cfg(test)]
+mod tests;
+
+#[path = "benchmarks.rs"]
+#[cfg(all(test, feature = "bench_private"))]
+mod benchmarks;
+
+const V2_COOKIE_BASE: u32 = 0x1c849303;
+
+const V2_COOKIE: u32 = V2_COOKIE_BASE | 0x10;
+
+const V2_HEADER_SIZE: usize = 40;
+
+/// Errors that occur during serialization.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum SerializeError {
+    /// A count above i64::max_value() cannot be zig-zag encoded, and therefore cannot be
+    /// serialized.
+    CountNotSerializable,
+    /// Internal calculations cannot be represented in `usize`. Use smaller histograms or beefier
+    /// hardware.
+    UsizeTypeTooSmall,
+    /// An i/o operation failed.
+    IoError
```

Comment:
Good news; the cost of including ErrorKind was unmeasurably small, probably because this format is not as optimizable as base64 so it just gets lost in the noise.